flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/sources/LookupableTableSource.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.sources;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.functions.AsyncTableFunction;
+import org.apache.flink.table.functions.TableFunction;
+
+/**
+ * A {@link TableSource} which supports for lookup accessing via key column(s).
+ * For example, MySQL TableSource can implement this interface to support lookup accessing.
+ * When temporal join this MySQL table, the runtime behavior could be in a lookup fashion.
+ *
+ * @param <T> type of the result
+ */
+@PublicEvolving
+public interface LookupableTableSource<T> extends TableSource<T> {
```

Comment:
What's the relationship between this `LookupableTableSource` and `DefinedIndexes` and `DefinedPrimaryKey`? Does this table source support getting lookup function for random keys, even if the columns have no index on them or not the primary key?

Reply:
Currently, the lookup key of `LookupableTableSource` must be either primary key or index. Therefore, a concrete `LookupableTableSource` should also implement `DefinedIndexes` or `DefinedPrimaryKey`.

> Does this table source support getting lookup function for random keys, even if the columns have no index on them or not the primary key?

No. It will throw exceptions.

Reply:
Do we have any better solution for this? It's hard for user to realize that `LookupableTableSource` should be used with `DefinedIndexes` or `DefinedPrimaryKey`

Reply:
As discussed offline, we will keep `DefinedIndexes` and `DefinedPrimaryKey` in the first version. And integrate them into table source v2 interface in the future.